include/dlaf/permutations/general/impl.h

```diff
@@ -152,4 +157,345 @@ void Permutations<B, D, T, C>::call(SizeType i_begin, SizeType i_end, Matrix<con
+    std::exclusive_scan(recv_counts.begin(), recv_counts.end(), recv_displs.begin(), 0);
+
+    // All-to-All communication
+    DLAF_MPI_CHECK_ERROR(MPI_Ialltoallv(send_ptr, send_counts.data(), send_displs.data(), dtype,
+                                        recv_ptr, recv_counts.data(), recv_displs.data(), dtype, comm,
+                                        req));
+  };
+
+  auto sender = ex::when_all(std::forward<SendCountsSender>(send_counts_sender),
+                             whenAllReadWriteTilesArray(send_mat),
+                             std::forward<RecvCountsSender>(recv_counts_sender),
+                             whenAllReadWriteTilesArray(recv_mat));
+  dlaf::comm::internal::transformMPIDetach(std::move(all2all_f), std::move(sender));
+}
+
+// @param nranks number of ranks
+// @param loc2gl_index  a column matrix that represents a map from local to global indices
+// @param packing_index a column matrix that represents a map from packed indices to local indices
+//        that is used for packing columns or rows to each rank
+//
+// Note: the order of the packed rows or columns on the send side must match the expected order at
+// unpacking on the receive side
+template <Device D, Coord C, bool PackBasedOnGlobalIndex>
+auto initPackingIndex(int nranks, SizeType i_el_begin, const matrix::Distribution& dist,
+                      Matrix<const SizeType, D>& loc2gl_index, Matrix<SizeType, D>& packing_index) {
+  namespace ex = pika::execution::experimental;
+  namespace di = dlaf::internal;
+
+  auto counts_fn = [nranks, i_el_begin, dist,
+                    len = packing_index.size().rows()](const auto& loc2gl_index_tiles,
+                                                       const auto& packing_index_tiles) {
+    const SizeType* in = loc2gl_index_tiles[0].get().ptr();
+    SizeType* out = packing_index_tiles[0].ptr();
+
+    std::vector<int> counts(to_sizet(nranks));
+    int offset = 0;
+    for (int rank = 0; rank < nranks; ++rank) {
+      int& count = counts[to_sizet(rank)];
+      count = 0;
+      for (SizeType i = 0; i < len; ++i) {
+        if (dist.rankGlobalElement<C>(i_el_begin + in[i]) == rank) {
+          if constexpr (PackBasedOnGlobalIndex) {
+            out[offset + count] = i;
+          }
+          else {
+            out[i] = offset + count;
+          }
+          ++count;
+        }
+      }
+      if constexpr (PackBasedOnGlobalIndex) {
+        std::sort(out + offset, out + offset + count,
+                  [in](SizeType i1, SizeType i2) { return in[i1] < in[i2]; });
+      }
+      offset += count;
+    }
+    return counts;
+  };
+
+  auto sender =
+      ex::when_all(whenAllReadOnlyTilesArray(loc2gl_index), whenAllReadWriteTilesArray(packing_index));
+  return ex::ensure_started(
+      di::transform<di::TransformDispatchType::Plain, false>(di::Policy<DefaultBackend_v<D>>{},
+                                                             std::move(counts_fn), std::move(sender)));
```

Comment:
```suggestion
  return ex::ensure_started(
      di::transform<di::TransformDispatchType::Plain, false>(di::Policy<Backend::MC>{},
                                                             std::move(counts_fn), std::move(sender)));
```
or you'll have to add a parameter pack on the lambda (and possibly an `if constexpr` to disable the body for GPU).

Reply:
I removed `Device D` as the function currently only supports `Device::GPU`. https://github.com/eth-cscs/DLA-Future/pull/676/commits/1150fae0afcb62d9b217f3feaf67ba07c3fede9d